rust/user-info-fetcher/src/util.rs

```diff
@@ -1,8 +1,54 @@
-use reqwest::RequestBuilder;
+use hyper::StatusCode;
+use reqwest::{RequestBuilder, Response};
 use serde::de::DeserializeOwned;
+use snafu::{ResultExt, Snafu};
 
-pub async fn send_json_request<T: DeserializeOwned>(
-    req: RequestBuilder,
-) -> Result<T, reqwest::Error> {
-    req.send().await?.error_for_status()?.json().await
+#[derive(Snafu, Debug)]
+pub enum Error {
+    #[snafu(display("failed to execute request"))]
+    HttpRequest { source: reqwest::Error },
+
+    #[snafu(display("failed to parse json response"))]
+    ParseJson { source: reqwest::Error },
+
+    #[snafu(display("response was an HTTP error: {text}"))]
+    HttpErrorResponse { status: StatusCode, text: String },
+
+    #[snafu(display("response was an HTTP error with undecodable text"))]
+    HttpErrorResponseUndecodableText {
+        status: StatusCode,
+        encoding_error: reqwest::Error,
+    },
+}
+
+pub async fn send_json_request<T: DeserializeOwned>(req: RequestBuilder) -> Result<T, Error> {
+    // make the request
+    let response = req.send().await.context(HttpRequestSnafu)?;
+    // check for client or server errors
+    let non_error_response = get_non_error_response(response).await?;
+    // parse the result
+    let result = non_error_response.json().await.context(ParseJsonSnafu)?;
+    Ok(result)
+}
+
+/// takes a Response and checks whether it is an error. If so, parse the reqwest Error
+/// and create our own error type with more context added. We do this because the plain
+/// reqwest error does not give any response body context.
```

Comment:
```suggestion
/// Wraps a Response into a Result. If there is an HTTP Client or Server error, extract the HTTP body (if possible) to be used as context in the returned Err. 
/// This is done this because the `Response::error_for_status()` method Err variant does not contain this information.
```

Reply:
_sorry, the suggestion feature and line breaks conflict_

Reply:
done: https://github.com/stackabletech/opa-operator/pull/510/commits/e0a5b9284263cf8cea8a7ceb6e1b5a8aaa4be8e8